### PR TITLE
Devise for user access

### PR DESCRIPTION
**Pull request by GeorgeJLewis**
### Devise for user access

###PT Link https://www.pivotaltracker.com/story/show/157767329

###What we did:
- Add devise gem
- Add signup, sign in to navbar
- Generate user model
- Add happy and sad path for restricting access
- Add Rspec test for user model

### CODE REVIEW

File: features/step_definitions/assertion_steps.rb [hunk 1]
```diff
@@ -1,3 +1,11 @@
 Then("I should see {string}") do |expected_text|
   expect(page).to have_content expected_text
 end
+
+Then("I should not see {string}") do |expected_text|
+  expect(page).to have_no_content expected_text
+end
+
+Given("I should be redirected to login page") do
+  expect(current_path).to include "sign_in"
```
Comment (tochman) [inline features/step_definitions/assertion_steps.rb]: Interesting solution but not how we usually do it. Please use the path helper

File: features/create_article_form.feature [hunk 2]
```diff
@@ -5,6 +5,7 @@ Feature: Article creation form
 
 Background:
   Given I am on the landing page
+  And user is signed in
```
Comment (tochman) [inline features/create_article_form.feature]: This feature is about the show page. Restrictions to creat is out of of scope

File: features/devise_for_user_access.feature [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+  When I click "The awesome article"
+  And I should see "The awesome article"
+  And I should see "Lorem ipsum dolor sit amet, consectetur adipisicing elit"
+  And I should not see "Another awesome articles"
+  And I should not see "Excepteur sint occaecat cupidatat non proident"
+
+Scenario: A user not logged in cannot access article show view
+  When I click "The awesome article"
+  And I should not see "Lorem ipsum dolor sit amet, consectetur adipisicing elit"
+  And I should be redirected to login page
```
Comment (tochman) [inline features/devise_for_user_access.feature]: Please check for a flash message

File: features/devise_for_user_access.feature [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+Background:
+  Given we have the following articles
+    | headline                  | content     |
+    | The awesome article       | Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. |
+    | Another awesome articles  | Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum. |
+
+  And I am on the landing page
+
+Scenario: A logged i user can access article show view
+  Given user is signed in
```
Comment (tochman) [inline features/devise_for_user_access.feature]: I usually try to login the user before you navigate to the site

File: features/article_have_view_page.feature [hunk 5]
```diff
@@ -8,6 +8,8 @@ Background:
     | headline                  | content     |
     | The awesome article       | Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. |
 
+  And user is signed in
+
```
Comment (tochman) [inline features/article_have_view_page.feature]: I see you use the role Subscriber in the scenario below. (I know this is legacy). Please revisit the scenario title and adjust it or create a PT ticket for it.

File: features/devise_for_user_access.feature [hunk 6]
```diff
@@ -0,0 +1,25 @@
+Feature: Logged in user can access article show view
+  As a user
+  In order to gain access to different features in the site
+  I should be able to create an account and login
```
Comment (tochman) [inline features/devise_for_user_access.feature]: Please use the user story I wrote yesterday. This one is simply not entirely relevant

**Review by tochman — CHANGES_REQUESTED** [review]

File: features/create_article_form.feature (hunk 2 above)
Comment (ghost) [reply features/create_article_form.feature]: This is a feature file created earlier. We needed to modify it to make it pass since we added a user model.

**Review by koggs — APPROVED** [review]
Looks good to me

**Review by alghazu — APPROVED** [review]
:+1:

**Comment by tochman** [comment]
The title does not reflect what the story is about

**Review by tochman — APPROVED** [review]
I will accept this but the way this feature has been handled shows that you guys still have a lot to learn and practice in terms of the Acceptance - Unit test cycle

**Pull request merged**